### PR TITLE
[not ready] Buildsystem Improvements

### PR DESCRIPTION
**Pull request by daniel-j-h**
This Pull Request tracks the progress. Do not merge yet.

The high level goal is to improve the build system in regard to warnings, build times, compiler and linker specific settings and features.

Please read every commit's description, as they have detailed explanations attached. This pull request just tracks the overall progress.

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Broken on [gcc 4.9.2](https://gist.github.com/daniel-j-h/512f62e908cf709fa395) --- see respective line in [car.lua](https://github.com/Project-OSRM/osrm-backend/blob/4062b6373b2bd218a22a2e7e44827e60ae8254e2/profiles/car.lua#L405).

**Comment by TheMarex** [comment]
Any open problems here? Can we merge?

File: CMakeLists.txt [hunk 1]
```diff
@@ -184,6 +190,28 @@ elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL "MSVC")
...
+    set(LINKER_FLAGS "${LINKER_FLAGS} -Wl,--gc-sections")
+  endif()
+  # Default linker optimization flags
+  set(LINKER_FLAGS "${LINKER_FLAGS} -Wl,-O1 -Wl,--hash-style=gnu -Wl,--sort-common")
+else()
+  message(STATUS "Using unknown linker, not setting linker optimizations")
+endif ()
+set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${LINKER_FLAGS}")
+# XXX(daniel-j-h): building libosrm.a invokes AR passing on the linker flags, but AR does not understand them
+#set(CMAKE_STATIC_LINKER_FLAGS "${CMAKE_STATIC_LINKER_FLAGS} ${LINKER_FLAGS}")
```
Comment (daniel-j-h) [inline CMakeLists.txt]: @TheMarex would love to fix this issue ^ before merging (not sure yet how to fix it)

**Comment by daniel-j-h** [comment]
Another pain point: the luabind checks that compile a small snippet that includes lua and luabind headers and links to the respective libs (hint: expensive) are [not cached](https://github.com/Project-OSRM/osrm-backend/blob/2143dc97fa65c786c6566db1343ad270347dcfd3/cmake/check_luabind.cmake#L2-L3). Why?

**Comment by daniel-j-h** [comment]
There are some warning flags which might help:

```
cmake .. -DCMAKE_BUILD_TYPE=Release -Wdev --warn-uninitialized --warn-unused-vars 
```
but the [log is somewhat verbose](https://gist.github.com/daniel-j-h/3e7e903c9b438a9eb23e) (but still finds e.g. `LTO_FLAGS` that is unused)

**Comment by daniel-j-h** [comment]
```
$ cmake .. -DCMAKE_BUILD_TYPE=Release --graphviz=deps.dot

$ for dot in (ls deps.dot*)
       dot -Tpng $dot -o $dot.png
   end
```
deps dot (click for high-res)

**Comment by daniel-j-h** [comment]
Would love to get [debug fission](https://gcc.gnu.org/wiki/DebugFission) to work, but for this we would first have to detect the `ld.gold` linker's availability.

**Comment by daniel-j-h** [comment]
`cmake/GetGitRevisionDescription.cmake` is broken.

We derive `libosrm`'s version by using the `git describe` command (e.g. on the commit hash above ^):

```
git describe a71066a
v4.5.0-625-ga71066a
```

This then is put into the `pkg-config` configuration file `libosrm.pc`.

**Comment by daniel-j-h** [comment]
The `pkg-config` configuration file `libosrm.pc` does not contain libraries for Intel TBB

This is the generated `.pc` config for both the static and shared `libosrm` target:

```
prefix=/usr/local
includedir=${prefix}/include/osrm
libdir=${prefix}/lib

Name: libOSRM
Description: Project OSRM library
Version: v4.5.0-625-ga71066a
Requires:
Libs: -L${libdir} -lOSRM
Libs.private: -L/usr/lib/x86_64-linux-gnu -lboost_date_time -lboost_filesystem -lboost_iostreams -lboost_program_options -lboost_regex -lboost_system -lboost_thread -lboost_unit_test_framework -lpthread
Cflags: -I${includedir}
```

**Comment by daniel-j-h** [comment]
Open issues here:

- [ ] version is taken from last git tag on _develop_, resulting in `v4.5.0` appearig in all our releases
- [ ] `pkg-config` does not correctly specify all required libs (tbb)

**Comment by freenerd** [comment]
I tested building this branch under mac and it worked. Now you know.

**Comment by daniel-j-h** [comment]
Building the shared library with lto fails. We clearly violate the ODR rule:

    cmake .. -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=1
    cmake --build .`

```
CMakeFiles/COORDINATE.dir/data_structures/coordinate.cpp.o (symbol from plugin): In function `FixedPointCoordinate::is_valid() const':
(.text+0x0): multiple definition of `FixedPointCoordinate::is_valid() const'
...
CMakeFiles/COORDINATE.dir/algorithms/coordinate_calculation.cpp.o (symbol from plugin): In function `coordinate_calculation::great_circle_distance(int, int, int, int)':
(.text+0x0): multiple definition of `coordinate_calculation::great_circle_distance(int, int, int, int)'

**Comment by daniel-j-h** [comment]
All done here!

@TheMarex in case you want to merge this in before `v4.8.0` note that the version is now taken from the `CMakeLists.txt` version variables. I already changed it to `4.8.0` in this branch.

**Comment by TheMarex** [comment]
@daniel-j-h what is blocking here? `node-osrm`? Otherwise I'm :+1: on merging here.

**Comment by daniel-j-h** [comment]
Done. Merging into develop as explained in https://github.com/Project-OSRM/node-osrm/issues/111#issuecomment-144463062.

**Pull request merged**